### DISCUSSION TITLE
Modelagem dos microsserviços com abordagem DDD

### DISCUSSION THREAD

**Discussion by ecureuill** [ecureuill/ADA_1042_AQ1_SB_CartaoService#39] [Ideas]
Passos necessários:
Definir
- [ ] Domínio
  -  Usuário
  - Compra
  - Cartão de Crédito
  - Fatura?
- [ ] Agregados
  - Usuário: entidades Usuário, Dependente e Cartão de Crédito
  - ...
- [ ] Regras de negócio
  - Um usuário só pode ter 1 cartão de crédito principal, mas pode ter vário cartões de créditos para seus dependentes  
- [ ] Comandos
  - Criar Usuário
  - Criar Cartão para dependente
  - Registrar Compra
  -  Consultar Fatura
- [ ] Eventos
  -  Usuário criado
  - Cartão de crédito criado
  - Compra registrada
  - Fatura consultada 

Idéia inicial

  - Microsserviço de Usuários:
        Responsável por lidar com a criação, atualização e consulta de usuários.
        Pode ter entidades como "Usuário" e "Dependente".

  -  Microsserviço de Cartões de Crédito:
        Responsável por criar, atualizar e consultar cartões de crédito.
        Pode ter entidades como "Cartão de Crédito" e "Dependente do Cartão".

  - Microsserviço de Compras:
        Responsável por registrar compras feitas com os cartões de crédito.
        Pode ter entidades como "Compra" e "Item de Compra".

  - Microsserviço de Faturas (?)
        Responsável por consultar faturas associadas aos cartões de crédito.
        Pode ter entidades como "Fatura" e "Item de Fatura".

**Comment by ecureuill** [comment]
https://learn.microsoft.com/pt-br/azure/architecture/microservices/model/domain-analysis

**Comment by ThusSpokeBieu** [comment]
Estabelecendo alguns detalhes do domínio, adicionando também algumas coisas que o @thiago-granja  desenhou ontem e outra coisas que eu pensei.

Podemos enxugar algumas regras de negócio caso fique um sistema muito complexo ou extenso, mas:

# Domínio: 

###  **Usuário**
- Acessa e interage com o sistema. 
- Quem recebe a fatura. 
- Possui um limite de compra/score. 
- Deve ser avisado sempre que um evento ocorre (abertura de conta, compra, fechamento de fatura e pagamento)

### **Dependentes**
- Estão dentro do guarda-chuva do usuário, mas não acessam o sistema diretamente. 
- Possuem cartão de crédito individuais, com limite compartilhado com o usuário.
- Suas compras são registradas no sistema e estão na fatura do usuário.

### **Cartão de Crédito**
- São confeccionados na abertura da conta do usuário, quando ele verifica se existe score o suficiente. 
- São enviados fisicamente ao usuário. 
- Podem ser cartão de crédito de dependente ou de usuário.

### Compra
- Deve registrar detalhes da compra, como estabelecimento onde ocorreu, setor, valor, horário e etc.
- Deve sempre verificar se existe score disponível e se o cartão está com credenciais corretas.
- Deve ser assincrona e reativa para evitar perdas de dados. 
- É parte crítica do sistema.

### Fatura
- É o registro de compras de um mesmo guarda-chuva (usuário e seus dependentes) dentro de um período de tempo.
- Não pode ser maior que o limite do usuário. 
- Possui sempre data de fechamento e uma data de vencimento.
- Deve-se adicionar juros quando passa da data de vencimento.
- Pode ocasionar em desativação de todos os cartões do guarda-chuva caso seja vencido por muito tempo.
- Também deve ser assincrona e reativa, visto que é outra parte crítica do sistema.

### Notificação
- Emite notificações de eventos ao usuário (e-mail e/ou sms).

**Comment by brunaminusso** [comment]
Construção dos diagramas:
https://drive.google.com/file/d/1f9EuABFcyFfCjoJf46CBzb6EEa9kD9rX/view?usp=sharing

**Reply by ecureuill** [reply]
Libera o acesso para colaboração :)

**Comment by ecureuill** [comment]
![image](https://github.com/ecureuill/ADA_1042_AQ1_SB_CartaoService/assets/993369/ec18eb72-6fb9-45aa-9064-27c72d0df0f3)

Estou pensando nessa alternativa: Microsserviço Financeiro (cartão + fatura)

**Comment by ecureuill** [comment]
Kanbam criado

https://github.com/users/ecureuill/projects/12/
